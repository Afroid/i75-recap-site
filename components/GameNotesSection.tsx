import React from "react";
import ReactMarkdown from "react-markdown";
import Image from "next/image";
import { GNSSection as GNS } from "@/types/types";

interface Props {
  data: GNS;
}

export default function GameNotesSection({ data }: Props) {
  return (
    <div>
      {data.pointsLeader && (
        <>
          <h3 className="text-xl font-semibold mt-2 mb-1">
            Points Leader: {data.pointsLeader.name}
            <span className="ml-1 text-green-500">
              ({data.pointsLeader.points})
            </span>
          </h3>
          {data.pointsLeader.notes && (
            <ReactMarkdown className="text-gray-800 mb-4">
              {data.pointsLeader.notes}
            </ReactMarkdown>
          )}
        </>
      )}

      {data.dud && (
        <>
          <h3 className="text-xl font-semibold mt-2 mb-1">
            Dud: {data.dud.name}
            <span className="ml-1 text-red-500">
              ({data.dud.points})
            </span>
          </h3>
          {data.dud.notes && (
            <ReactMarkdown className="text-gray-800 mb-4">
              {data.dud.notes}
            </ReactMarkdown>
          )}
        </>
      )}

      {Array.isArray(data.matchups) && data.matchups.map((matchup, i) => (
        <div key={i} className="border rounded p-4 mb-6 bg-gray-50">
          <p className="text-md font-medium mb-1">
            {matchup.team1} vs {matchup.team2} â€”
            <span className="font-semibold">{matchup.winner} wins</span>
          </p>
          <p className="text-sm text-gray-600 mb-2">{matchup.score}</p>
          <ReactMarkdown className="text-gray-800 whitespace-pre-line mb-2">
            {matchup.breakdown}
          </ReactMarkdown>
          {matchup.imageUrl && (
            <Image
              src={matchup.imageUrl}
              alt={matchup.imageAlt || "Matchup image"}
              width={600}
              height={400}
              className="rounded shadow-md"
            />
          )}
        </div>
      ))}
    </div>
  );
}
