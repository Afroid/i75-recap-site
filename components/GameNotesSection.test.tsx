import React from "react";
import { render, screen } from "@testing-library/react";
import type { GNSSection as GNS } from "@/types/types";
import type { ScoreEntry } from "./Scoreboard";

// Stub Scoreboard so it just renders the entries JSON
jest.mock("./Scoreboard", () => ({
  __esModule: true,
  default: ({ entries }: { entries: ScoreEntry[] }) => (
    <div data-testid="game-notes-scoreboard">
      {JSON.stringify(entries)}
    </div>
  ),
}));

// Stub ReactMarkdown to render its children directly
jest.mock("react-markdown", () => (props: { children: string }) => (
  <div data-testid="stubbed-game-notes-breakdown">
    {props.children}
  </div>
));

// Stub next/image to a simple <img>
jest.mock("next/image", () => ({
  __esModule: true,
  default: (props: React.ImgHTMLAttributes<HTMLImageElement>) => (
    <img data-testid="game-notes-image" {...props} />
  ),
}));

/**
 * Importing after the mocks ensures the real copies of those dependencies don't get loaded.
 * This ensures any imports from next/head or whatever get routed to the stubbed versions, rather
 * than real implementation. This makes sure jest.mock calls will have effect.
 */
// Now import GameNotesSection (after the mocks)
import GameNotesSection from "./GameNotesSection";

describe("GameNotesSection", () => {
  const baseMatchup = {
    team1:      "Player X",
    team1Logo:  "/playerX.png",
    record1:    "3-1",
    score1:     21,
    team2:      "Player Z",
    team2Logo:  "/playerZ.png",
    record2:    "2-2",
    score2:     14,
    breakdown:  "Player X dominated from the start.",
  };

  it("renders entries, breakdown, and image when imageUrl is present", () => {
    const data: GNS = {
      type: "gameNotes",
      matchups: [
        { ...baseMatchup, imageUrl: "/highlight.gif" }
      ]
    };

    render(<GameNotesSection data={data} />);

    // The main root container
    const container = screen.getByTestId("game-notes-section");
    expect(container).toBeInTheDocument();

    // Scoreboard entries stub
    const scoreboard = screen.getByTestId("game-notes-scoreboard");
    expect(scoreboard).toHaveTextContent(
      JSON.stringify([
        {
          logoUrl: "/playerX.png",
          name:    "Player X",
          record:  "3-1",
          score:   21,
        },
        {
          logoUrl: "/playerZ.png",
          name:    "Player Z",
          record:  "2-2",
          score:   14,
        },
      ])
    );

    // Breakdown text
    expect(screen.getByTestId("stubbed-game-notes-breakdown")).toHaveTextContent(
      "Player X dominated from the start."
    );

    // Image container & its image
    const img = screen.getByTestId("game-notes-image") as HTMLImageElement;
    expect(img.src).toContain("/highlight.gif");
    expect(img.alt).toBe("Player X vs Player Z highlight");
  });

  it("renders entries with default logos, no records, breakdown, and an image", () => {
    const baseMatchup = {
      team1:      "Player X",
      team1Logo:  "",
      record1:    "",
      score1:     21,
      team2:      "Player Z",
      team2Logo:  "",
      record2:    "",
      score2:     14,
      breakdown:  "Player X dominated from the start.",
    };

    const data: GNS = {
      type: "gameNotes",
      matchups: [
        { ...baseMatchup, imageUrl: "/highlight.gif" }
      ]
    };

    render(<GameNotesSection data={data} />);

    // The main root container
    const container = screen.getByTestId("game-notes-section");
    expect(container).toBeInTheDocument();

    // Scoreboard entries stub
    const scoreboard = screen.getByTestId("game-notes-scoreboard");
    expect(scoreboard).toHaveTextContent(
      JSON.stringify([
        {
          logoUrl: "/default-RWB.png",
          name:    "Player X",
          record:  "",
          score:   21,
        },
        {
          logoUrl: "/default-logo-alt.png",
          name:    "Player Z",
          record:  "",
          score:   14,
        },
      ])
    );

    // Breakdown text
    expect(screen.getByTestId("stubbed-game-notes-breakdown")).toHaveTextContent(
      "Player X dominated from the start."
    );

    // Image container & its image
    const img = screen.getByTestId("game-notes-image") as HTMLImageElement;
    expect(img.src).toContain("/highlight.gif");
    expect(img.alt).toBe("Player X vs Player Z highlight");
  });

  it("omits the image block when imageUrl is undefined", () => {
    const data: GNS = {
      type: "gameNotes",
      matchups: [
        { ...baseMatchup /* no imageUrl */ }
      ]
    };

    render(<GameNotesSection data={data} />);

    // Asertions
    expect(screen.getByTestId("game-notes-scoreboard")).toBeInTheDocument();
    // No image was rendered
    expect(screen.queryByTestId("game-notes-image")).toBeNull();
  });

  it("matches the snapshot for markup structure", () => {
    const data: GNS = {
      type: "gameNotes",
      matchups: [
        { ...baseMatchup, imageUrl: "/gif.gif" }
      ]
    };
    const { asFragment } = render(<GameNotesSection data={data} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
