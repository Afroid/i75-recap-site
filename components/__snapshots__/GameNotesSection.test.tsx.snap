// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameNotesSection matches the snapshot for markup structure 1`] = `
<DocumentFragment>
  <div
    data-testid="game-notes-section"
  >
    <div
      class="mb-6"
      data-testid="game-notes-item-0"
    >
      <div
        data-testid="game-notes-scoreboard"
      >
        [{"logoUrl":"/playerX.png","name":"Player X","record":"3-1","score":21},{"logoUrl":"/playerZ.png","name":"Player Z","record":"2-2","score":14}]
      </div>
      <div
        class="mt-2 text-gray-800"
        data-testid="game-notes-breakdown"
      >
        <div
          data-testid="stubbed-game-notes-breakdown"
        >
          Player X dominated from the start.
        </div>
      </div>
      <div
        class="mt-4"
        data-testid="game-notes-image-container"
      >
        <img
          alt="Player X vs Player Z highlight"
          class="w-full lg:w-auto h-auto rounded shadow-md"
          data-testid="game-notes-image"
          height="400"
          src="/gif.gif"
          width="600"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
